### ISSUE TITLE
[API Proposal]: Get default value of a `Type`

### ISSUE THREAD

**Issue by Neme12** [dotnet/runtime#90697]
### Background and motivation

When I have a `System.Type` and want to get its default value, the usual answer has always been to use code like this:
```c#
public static object? GetDefault(Type type)
{
   if (type.IsValueType)
   {
      return Activator.CreateInstance(type);
   }
   return null;
}
```
e.g. see https://stackoverflow.com/questions/325426/programmatic-equivalent-of-defaulttype

However, C# now allows explicit parameterless constructors on structs, so `new T()` can now be different from `default(T)` for structs, and therefore the code above  is incorrect. It's also not clear what a proper way to do this would look like and it doesn't seem to be trivial at all. That's why I'm proposing an API for this.

### API Proposal

```csharp
namespace System
{
    public abstract partial class Type
    {
        public virtual object? DefaultValue { get; }
        // or
        public virtual object? GetDefaultValue();
    }
}
```

### API Usage

.

**Comment by Neme12** [comment]
I *think* the right way to do this is now to use `RuntimeHelpers.GetUninitializedObject`:
```c#
public static object? GetDefault(Type type)
{
   if (type.IsValueType)
   {
      return RuntimeHelpers.GetUninitializedObject(type);
   }
   return null;
}
```

**Comment by steveharter** [comment]
> I think to do this correctly you now have to use RuntimeHelpers.GetUninitializedObject:

Yes `GetUninitializedObject()` is the correct API to create a value type with default values and without running any constructor which would be the same as `default(T)`.

**Comment by steveharter** [comment]
Adding to v9 for consideration - making it easier to obtain the default value of a value type given only a late-bound `Type` instance; it would have the same semantics as calling the compile-time `default(T)`.

**Comment by Neme12** [comment]
@steveharter 
![image](https://github.com/dotnet/runtime/assets/11444821/49705620-882a-4d9e-9ad9-d923ca4df158)
I think the API should work for reference types as well for consistency, not just for value types

**Comment by Neme12** [comment]
On the other hand, there's an existing `ParameterInfo.DefaultValue` property and this seems to be similar to that. In both cases the value is a compile-time constant.

**Comment by MichalPetryka** [comment]
> * I think a static method on `System.Activator` would be more appropriate place for this API.

Why not RuntimeHelpers?



> * The API should be called `CreateDefaultValue` (assuming that it is not expected to cache the instance).

How would the API be supposed to cache instances here? With reference types it'll always be just null and with structs it'll just be zero bytes that'll be returned by value so caching doesn't make sense in this context?

**Comment by Neme12** [comment]
If this was on `Type`, it would be trivial to cache the value on the instance. On `Activator`, I don't see how it could be cached :(

**Comment by Neme12** [comment]
I think it would definitely make sense to cache this.

**Comment by Neme12** [comment]
Another reason I would prefer this on `Type` rather than `Activator` is that today, everything you can do with `Activator`, you can also do explicitly by calling a specific constructor. I never use `Activator` personally and always explicitly find the constructor I'm interested in invoking and then invoke it. But with this API, there wouldn't be an alternative way of doing this and I would now have to use `Activator` whereas today, I can avoid it completely.

**Comment by steveharter** [comment]
Caching at some level is necessarily if this is expected to be called frequently for the same type, as `GetUninitializedObject()` doesn't do any.

Adding a new field for caching purposes to System.Type will increase the memory size a bit for all type instances for this somewhat niche use.

@Neme12 what is your scenario? Typically for late-bound uses of reflection for serializers etc they would cache other various type-related info, which could also cache the results from `GetUninitializedObject()`.

Also were you planning to implement this?

**Comment by vladd** [comment]
> If this was on `Type`, it would be trivial to cache the value on the instance. On `Activator`, I don't see how it could be cached :(

Something like this ([sharplab.io link](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUH1TATgAoBBAY2AEsA3AQ2AHswA6AcQFNgARTgM3oQANsABq9YRE4AealGAA+EgEoVAbgLFyVOoxYdufQSPGTpMwgAZlazbgAEThwQDMDwgDYP6B5RoMzGB4AN54zh6Y3hgOxkKiElKcAMpB9ADmsgAqiuHOYY4RzqjuXg5gnPRwTFDCAJ4OWQD8DonSDgC8DnAC8cD2EQC+eU4jHqVRjS1cvL2mbdm2nYqxcwnmKWmZMjmsC/aDQA==))?

**Comment by OwnageIsMagic** [comment]
@Neme12 you do not need to specify values for `out` parameters. [sharplab.io](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAIBYAEARApgYwHsoBnYSfYTACkIioCNDCAbTACwEMSA1TliLhiY6VAJZQqAN36CAlJgDeHbnwG5MAXkzlBAbkwz1WzAE4DAXwCwAKBlhMcE3iKldlPbfuZOYB9qhcAHcRBgArAmAAbQAmAF1PGzgAOgBZXGB2QhSASSgpQgBrXGooCBYWYV8wOUTUAEZkhtNqaswxEkwo3SEzONrMAHpBnUhcIA==)
```cs
void Deconstruct (out bool hasValue, out int value) { hasValue = true; value = 9; }
var d = Deconstruct;
var arr = new object[2];
d.Method.Invoke(null, arr);
System.Console.WriteLine(arr is [true, 9]); // true
```
@jkotas I don't think it's not ok to cache values, since it may be mutable struct and you can `Unsafe.Unbox` it.
`ActivatorCache` used by `Activator.CreateInstance` already have `CreateUninitializedObject`, so it should be easy to just expose it as `Activator.CreateDefaultValue`.
Done some benchmarks https://gist.github.com/OwnageIsMagic/821d31dbd575f1d207ac15a385a533f4
Currently `RuntimeHelpers.GetUninitializedObject` is around 4x slower than `Activator.CreateInstance`.
```
BenchmarkDotNet v0.13.12, Windows 10 (10.0.19045.3448/22H2/2022Update)
Intel Core i5-8500 CPU 3.00GHz (Coffee Lake), 1 CPU, 6 logical and 6 physical cores
.NET SDK 8.0.100
  [Host]     : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX2
  DefaultJob : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX2


| Method                  | ObjType      | Mean      | Error     | StdDev    | Ratio | RatioSD |
|------------------------ |------------- |----------:|----------:|----------:|------:|--------:|
| Direct                  | MyStruct     |  4.542 ns | 0.0425 ns | 0.0377 ns |  1.00 |    0.00 |
| DirectNew               | MyStruct     |  4.729 ns | 0.0459 ns | 0.0407 ns |  1.04 |    0.01 |
| ActivatorCreateInstance | MyStruct     | 10.728 ns | 0.0456 ns | 0.0426 ns |  2.36 |    0.02 |
| GetDefaultValueCommon   | MyStruct     | 12.823 ns | 0.0860 ns | 0.0718 ns |  2.82 |    0.03 |
| CreateDelegateCached    | MyStruct     | 16.085 ns | 0.0609 ns | 0.0509 ns |  3.54 |    0.03 |
| GetUninitializedObject  | MyStruct     | 41.614 ns | 0.1316 ns | 0.1231 ns |  9.16 |    0.09 |
|                         |              |           |           |           |       |         |
| Direct                  | System.Int32 |  4.409 ns | 0.0523 ns | 0.0489 ns |  1.00 |    0.00 |
| DirectNew               | System.Int32 |  4.734 ns | 0.0319 ns | 0.0283 ns |  1.07 |    0.01 |
| ActivatorCreateInstance | System.Int32 |  8.918 ns | 0.0328 ns | 0.0291 ns |  2.02 |    0.02 |
| GetDefaultValueCommon   | System.Int32 |  5.694 ns | 0.0729 ns | 0.0682 ns |  1.29 |    0.02 |
| CreateDelegateCached    | System.Int32 | 14.994 ns | 0.0465 ns | 0.0388 ns |  3.40 |    0.04 |
| GetUninitializedObject  | System.Int32 | 35.708 ns | 0.1412 ns | 0.1321 ns |  8.10 |    0.09 |
```

**Comment by jkotas** [comment]
> I don't think it's not ok to cache values, since it may be mutable struct and you can Unsafe.Unbox it.

Boxes structs are assumed to be immutable by the runtime. Mutating boxed struct is undefined behavior.

> Currently RuntimeHelpers.GetUninitializedObject is around 4x slower than Activator.CreateInstance

No longer the case in .NET 9. Fixed by #98016

**Comment by OwnageIsMagic** [comment]
https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.unsafe.unbox?view=net-8.0#remarks
> must not mutate the returned reference unless they are certain that T is a mutable struct type.

Remark hints that it's acceptable in some cases.

**Comment by MichalPetryka** [comment]
> > I don't think it's not ok to cache values, since it may be mutable struct and you can Unsafe.Unbox it.
> 
> Boxes structs are assumed to be immutable by the runtime. Mutating boxed struct is undefined behavior.

`Unbox` returns a "controlled-mutability managed pointer" which can be legally mutated with instance methods. You don't need to use Unbox however, casting a boxed struct to an interface it implements and calling a method there can also mutate the box without any unsafe code.

**Comment by jkotas** [comment]
Ah ok. You are right - this only applies to readonly structs.
